modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js

```diff
@@ -258,6 +272,12 @@ export default function AuditBarChart({rtl, vocabularies}) {
 							barSize={BAR_CHART.barHeight}
 							dataKey="value"
 							fill={COLORS[0]}
+							onMouseOut={() => {
+								setTooltip(null);
+							}}
+							onMouseOver={() => {
+								setTooltip('value');
```

Comment:
I don't fully understand what is going on here, so just a quick question:

Are we sure we want to use a magic string (`'value'`) as a cue to change the behavior of the `CustomToolTip`? It may seem unlikely, but what if `'value'` ever collides with a `dataKey` of the same name? eg. I am thinking of vocabularies like this one:

```
categories: [
	{key: 'growth', name: 'Growth', value: 478, vocabularyName: 'Stock type'},
	{key: 'income', name: 'Income', value: 1055, vocabularyName: 'Stock type'},
	{key: 'value', name: 'Value', value: 822, vocabularyName: 'Stock type'},
],
```

So on line 305 below, for this data, the key `'value'` would show `label` instead of `payload[i].name` (ie. "Value"). I don't actually know what `label` would be though...

Reply:
It's a little bit tricky because we have two options to display a `<Bar>`:

**1.** One when we have categories inside our vocabulary: two vocabularies selected in the configuration

```js
{bars.length &&
	bars.map((bar, index) => {
		return (
			<Bar
				barSize={
					checkboxes[bar.dataKey] === true
						? BAR_CHART.barHeight
						: 0
				}
				dataKey={bar.dataKey}
				fill={colors[bar.dataKey]}
				hide={checkboxes[bar.dataKey] !== true}
				key={index}
				legendType="square"
				name={bar.name}
				onMouseOut={() => {
					setTooltip(null);
				}}
				onMouseOver={() => {
					setTooltip(bar.dataKey);
				}}
			/>
		);
	})}
```
**2.** Another one when we don't have categories inside our vocabulary: one vocabulary selected in the configuration

```js
{!bars.length && (
	<Bar
		barSize={BAR_CHART.barHeight}
		dataKey="value"
		fill={COLORS[0]}
		onMouseOut={() => {
			setTooltip(null);
		}}
		onMouseOver={() => {
			setTooltip('value');
		}}
	/>
)}
```

Your concern goes with the first option that I mentioned and in that case, we use `bar.dataKey`. When there are no categories, the `datakey` is `"value"` (see point 2). But I think definitely we can refactor this to not hardcoded "value" in setTooltip state. I'll update the PR. Thanks!

Reply:
The `label` comes from the custom tooltip information via props. If you saw the code above, I'm using the same values as in `dataKey` to set the tooltip. I tried to refactor but if I pass `props.payload` to `setTooltip` more logic is required inside the `renderTooltip` method 🤔

Reply:
I don't know what Recharts' `<ToolTip />` will do if you pass it a non-primitive--string `tooltip` prop. It could be as simple as something like:

```js
// At the top-level of the module:

// Technically, still a string, and unlike 'value', it has a unique identity.
// ie. VALUE_TOOLTIP !== 'value'
const VALUE_TOOLTIP = new String('value');

// ... then in the component ...
setTooltip(VALUE_TOOLTIP) // instead of `setTooltip('value');`

// ... which means that on line 297 you can now do:
if (payload[i].dataKey == tooltip) {
  // Note the double-eq... not a huge fan of it, but...

  // and then on line 305
  tooltip === VALUE_TOOLTIP ? label : payload[i].name
  // note the triple-eq
```

I don't really like the subtlety of it — the reliance on the behavior of `===` vs `==` — or even the fact that the word "value" is somehow tied to the use of the *label*.

I think the simplest fix is instead just to choose a sentinel value will never (realistically) collide with a real value; eg.

```js
// At the top level of the module:

const USE_LABEL_AS_TOOLTIP = '__LABEL_AS_TOOLTIP__';

// Line 271-273 etc:
<Bar
    barSize={BAR_CHART.barHeight}
    dataKey={USE_LABEL_AS_TOOLTIP}
    ...
    onMouseOver(() => {
        setTooltip(USE_LABEL_AS_TOOLTIP);
    }}
>

// Line 305:
tooltip === USE_LABEL_AS_TOOLTIP ? label : payload[i].name
```

Reply:
To represent the value of the `Bar` the library uses `dataKey`. And that `dataKey` should exist in the `data` that we inject to the `<BarChart>` component as a key:

```js
<BarChart
	data={data}
	height={BAR_CHART.height}
	width={BAR_CHART.width}
>
...
</BarChart>
```

```
const data = [
    {key: 'business-decision-maker', name: 'Business Decision Maker', value: 987, vocabularyName: 'Audience'},
    {...},
]
```

So I think here we can either change as option 1 (use `VALUE_TOOLTIP`) or keep it as we have now.  What do you think?

Reply:
Sigh... Well, I don't like the fragility and subtlety of the "option 1" (`new String()` trick), but I already provided [a plausible example](https://github.com/liferay-frontend/liferay-portal/pull/80#discussion_r450773248) where the word "value" could be a totally legitimate example of a `dataKey`.

I still don't know what `label` actually looks like, though, because [your answer above](https://github.com/liferay-frontend/liferay-portal/pull/80#discussion_r450949325) only says:

> The `label` comes from the custom tooltip information via props.

How bad would it be if the ternary on line 305 ended up using `label` when it shouldn't?

Ultimately this is trade-off of handling an edge case vs complexity. I am happy with whichever side you think is the most important — the worst case scenario is about the same on both sides (ie. you end up with a future bug report because someone runs into the edge case, or you end up with a future bug report because the complexity led us to make a mistake somewhere).

Reply:
Given what we chatted about on Slack:

- React propType validation is going to complain about `new String()` (because it is obviously doing a `typeof blah === 'string'` check, and `typeof new String('blah')` is `'object'`); so we can rule out that approach.
- In the multiple bars case, there is no `payload[i].name` property, (instead it is under `payload[i].payload.name`, duplicated under `label`).
- So, we should be able to leverage that to decide whether to use `label` or not.

Does this work?

```diff
diff --git a/modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js b/modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js
index 2404167497c1..c3a65d87e7ba 100644
--- a/modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js
+++ b/modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js
@@ -302,7 +302,7 @@ function CustomTooltip(props) {
 					style={{width: 150}}
 				>
 					<ClayLayout.ContentCol expand>
-						{tooltip === 'value' ? label : payload[i].name}
+						{payload[i].name ? payload[i].name : label}
 					</ClayLayout.ContentCol>
 					<ClayLayout.ContentCol>
 						{payload[i].value}
```

If it doesn't, I'm inclined to accept defeat.

Aren't declarative APIs wonderful?

Reply:
That is not going to work because `payload[i].name` exists either if we have selected two vocabularies or just one. In case we have selected one, `payload[i].name` is going to be `value` (as the `dataKey`) and this change will cause:

<img width="1250" alt="Screenshot 2020-07-08 at 16 10 39" src="https://user-images.githubusercontent.com/15845466/86928981-95455d80-c135-11ea-875f-d9df63f3a223.png">

...and that's not what we want to display in the tooltip 😢 

I think we were defeated.

Reply:
Ok, well feel free to ship this. Given the partial screenshots of the data structure that you shared in Slack, I have run out of ideas.

Reply:
That makes me feel bad, but I'm out of ideas too 😢 Let's retest this. Thank you so much for your help and assistance to try to find another solution 👍

Reply:
Take a look at https://github.com/dgarciasarai/liferay-portal/pull/266 and see if that makes any sense. But like I said, fine with shipping this as-is (it is only an edge case, after all).